src/cargo/core/resolver/mod.rs

```diff
@@ -1064,28 +1162,33 @@ fn activate_deps_loop(
                         features = frame.features;
                         conflicting_activations = frame.conflicting_activations;
                         backtracked = true;
-                        (candidate, has_another)
-                    })
-                    .ok_or_else(|| {
-                        activation_error(
-                            &cx,
-                            registry.registry,
-                            &parent,
-                            &dep,
-                            &conflicting_activations,
-                            &candidates,
-                            config,
-                        )
-                    })
+                        Ok((candidate, has_another))
+                    }
+                    None => Err(activation_error(
+                        &cx,
+                        registry.registry,
+                        &parent,
+                        &dep,
+                        &conflicting_activations,
+                        &candidates,
+                        config,
+                    )),
+                }
             })?;
 
+            // TODO: explain this
             if just_here_for_the_error_messages && !backtracked && has_another {
```

Comment:
:laughing:

Reply:
We know we are going to fail all the way to the user: `if just_here_for_the_error_messages` and
yet we have another thing to try: `has_another` so if we try this it won't make it all the way to the user. Insted it will fail and bactrak to us to try the next candidate. So let's just skip to the next candidate.

Reply:
Aha that definitely makes sense! 

Put another way, if `just_here_for_the_error_messages` is true, we basically want to exhaust the iterator immediately to collect the `HashMap` it spits out as to why we filtered out candidates?

Reply:
Sort of. If `just_here_for_the_error_messages` then we ether want to 
1. Exhaust the iterator immediately (as you said) so this dep can bactrack to the users.
2. Activate the last candidate so one of our dependencies can bactrack to the users.

Reply:
Oh hm I'm not sure I understand then. For that second condition where we want to activate the last candidate so we can backtrack, how come the last candidate is special? Is that because it's the only one which'll force a backtrack if it fails?

Reply:
Yes? It is the only one that does not add a backtrack frame for our children to come back to. And we could just pick the first one and not add a backtrack frame, but then the error message will be different.

Befor:```    ... which is depended on by `bar v0.1.0` ```
After: ```    ... which is depended on by `bar v0.1.99` ```